WIP Tab groups

noctuid opened the pull request:
Addresses #811.

I will rebase/squash and update the title once the PR is finalized/ready. I have never worked with webextensions before and don't have much typescript experience (or any javascript experience). Any feedback is appreciated.

Functionality implemented
- `:tgroupcreate` to create a tab group
- `:tgroupswitch` to switch to or create a tab group
- `:tgrouplast` to switch to the previously active tab group
- `:tgrouprename` to rename the current tab group
- `:tgroupclose` to close all tabs in the current tab group
- `:tgroupmove` to move the current tab to another tab group
- `:tgroupabort` to destroy all tab group information and show all tabs
- Restoring tab group setup on firefox start (seems to just work without any extra code)
- Basic mode indicator
- Setting for initial urls for to use newly created tab groups (after first)
- Setting for whether :tab shows hidden tabs.

Edge cases handled
- Opening tabs outside of tridactyl mechanisms (e.g. firefox gui, C-t, etc.)
- Switching to a hidden tab
- Detaching and attaching tabs
- Pinned tabs (they can't be hidden; the commands ignore them)
- Various other edge cases; let me know if you find any I've missed

Functionality not yet implemented
- [ ] Mode indicator is not fully working (see question below)
- [ ] Edge case where last tab in a tab group is deleted; I thought I'd fixed this, but I've only been able to get it to work in the temporary profile firefox and not my own firefox

Functionality for later
- [ ] Completion for `:tgroupmove` and `:tgroupswitch` Is it okay if this is left for later? I would personally never use this functionality.
- [ ] `:tgroupget`, I never used this with pentadactyl (I never really used the `:tgroupmove` equivalent either)
- [ ] `:undo` completion restriction - Tab values seem to be preserved when restoring a tab. With the current code, it will switch back to a previous group if the last delete was on a different group. It doesn't try to restrict the undo completions. I don't mind this behavior and haven't looked into whether it would even be possible to restrict undo to the current tab group. Maybe a setting could be added later if it is possible and someone wants it.

Questions
- Where can I find the types to add for the parameters passed to tabs.onX functions (e.g. type for activeInfo for browser.tabs.onActivated)?
- It's not completely clear to me whether to use config.get or config.getAsync after reading the docstrings and then looking at how they are used.
- How can I get the mode indicator to update when things happen that change the tab group name? Right now it's using addContentStateChangedListener which doesn't detect changes immediately if I run something like `:tgroupcreate` or `:tgrouprename`.
- Probably the way I am using a bunch of "as string" and "as number"s is not great. Right now, validation is done in excmds and all helper functions assume they have been called correctly (e.g. a helper may require multiple tab groups to exist, but it doesn't check this itself). Where I've used "as string", it means that the value should be guaranteed to have been set (unless something has gone horribly wrong), but would it be better to throw errors if the something happens to be undefined or do something else instead?
- Given that printWidth is disabled, is there a line length I should try to limit things to?
- I seem to have broken tridactyl on all pages except the newtab page? Or is this an issue with my setup?

glacambre commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+
+    groups.add(name)
+    promises.push(setTgroups(groups))
+    return Promise.all(promises).then(() => undefined)
```
I think we could return `name` instead of `undefined` here. This would allow nice things like `:js tri.excmds.tgroupcreate("mygroup").then(n => /* do something with n rather than retyping "mygroup" */)`

noctuid replied:
I've updated all commands to return the name of the new tab group.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ *
+ */
+export async function windowTgroup(id?: number) {
+    return browserBg.sessions.getWindowValue(id || await activeWindowId(), "tridactyl-active-tgroup")
```
```suggestion
    return browserBg.sessions.getWindowValue(id || await activeWindowId(), "tridactyl-active-tgroup") as string
```
This would avoid the need for for multiple casts in other functions (e.g. tgroupActiveLast/tgroupHandleTabCreated).

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ *
+ */
+export async function tgroupActivateLast() {
+    const otherGroupsTabs = await tgroupRun(await windowTgroup() as string, true)
```
```suggestion
    const otherGroupsTabs = await tgroupRun(await windowTgroup(), true)
```
If you add the cast to the return value of windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ *
+ */
+export async function tgroupHandleTabCreated(tab: browser.tabs.Tab) {
+    const windowGroup = await windowTgroup() as string
```
```suggestion
    const windowGroup = await windowTgroup()
```
If you add the cast to windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+    // NOTE this doesn't need to worry about a tab on another window previously
+    // being pinned because tabs become unpinned when you move them between
+    // windows
+    const windowGroup = await windowTgroup() as string
```
```suggestion
    const windowGroup = await windowTgroup()
```
If you add the cas to windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ * show all its tabs, and hide all tabs from the old tab group.
+ */
+export async function tgroupHandleTabActivated(activeInfo) {
+    const windowGroup = await windowTgroup(activeInfo.windowId) as string
```
```suggestion
    const windowGroup = await windowTgroup(activeInfo.windowId)
```
If you add the cast to windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ */
+export async function tgroupHandleTabUpdated(tabId: number, changeInfo) {
+    if (changeInfo.pinned !== undefined) {
+        const windowGroup = await windowTgroup() as string
```
```suggestion
        const windowGroup = await windowTgroup()
```
If you add the cast to windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ */
+export async function tgroupHandleTabRemoved(_tabId: number, removeInfo) {
+    if (!removeInfo.isWindowClosing) {
+        const windowGroup = await windowTgroup() as string
```
```suggestion
        const windowGroup = await windowTgroup()
```
If you add the cast to windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ *
+ */
+export async function tgroupHandleTabDetached(tabId: number, detachInfo) {
+    const windowGroup = await windowTgroup() as string
```
```suggestion
    const windowGroup = await windowTgroup()
```
If you add the cast to windowTgroup.

glacambre commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+        throw new Error("No tab groups exist")
+    }
+
+    return tgroupClearOldInfo(await windowTgroup() as string, name).then(() => undefined)
```
```suggestion
    return tgroupClearOldInfo(await windowTgroup(), name).then(() => undefined)
```
If you put the cast in windowTgroup.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+    if (newName) {
+        promises.push(setWindowTgroup(newName))
+        promises.push(
+            tgroupRun(oldName).then(async tabs => {
+                await setTabTgroup(newName, tabs.map(tab => tab.id))
+            })
```
No need for async/await here.
```suggestion
           tgroupRun(oldName)
                .then(tabs => setTabTgroup(newName, tabs.map(tab => tab.id)))
                .then(() => undefined)
```

glacambre commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+    } else if (groups.size == 1) {
+        throw new Error("This is the only tab group")
+    } else if (groups.size > 1) {
+        const closeGroup = await windowTgroup() as string
```
```suggestion
        const closeGroup = await windowTgroup()
```
If you move the cast to windowTgroup.

glacambre commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+//#background
+export async function tgroupmove(name: string) {
+    const groups = await tgroups()
+    const currentGroup = await windowTgroup() as string
```
```suggestion
    const currentGroup = await windowTgroup()
```
If you move the cast to windowTgroup.

glacambre commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+    // switch to other group if this is the last tab in the current group
+    if (tabCount == 1) {
+        return Promise.all([
+            tgroupClearOldInfo(currentGroup),
+            setWindowTgroup(name),
+            setTabTgroup(name, currentTabId),
+            tgroupRun(name).then(tabs => browserBg.tabs.show(tabs.map(tab => tab.id))),
```
Wouldn't it be sufficient to call `tgroupClearOldInfo(currentGroup, name)` here? Or does this do something more?

noctuid replied:
Yes, updated. Still need to show the new tabs though.

glacambre commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+            tgroupRun(name).then(tabs => browserBg.tabs.show(tabs.map(tab => tab.id))),
+        ]).then(() => undefined)
+    } else {
+        const lastTabId = await tgroupLastTabId(currentGroup)
+        await tabSetActive(lastTabId)
+        return browser.tabs.hide(currentTabId).then(() => undefined)
```
I think there's something I don't understand here - as far as I can tell, this makes the currently active tab active and then hides it, without moving any tab anywhere. Is there something I'm missing?

noctuid replied:
It's activating the previous tab not the current tab. It's already set the tab group name to the group it's being moved to, so all it needs to do is hide the tab. `:tgroupmove` stays on the current tab group except when run for the last tab in the tab group.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+        clearWindowTgroup(),
+        browser.tabs.query({currentWindow: true}).then(async tabs => {
+            const ids = tabs.map(tab => tab.id)
+            await browser.tabs.show(ids)
```
Could there be tabs that don't belong to a tabgroup and that were hidden by another addon that would be made visible by this?

noctuid replied:
Any new tab is automatically added to the current tab group. I think trying to integrate with another addon that hides tabs would be difficult. How do we know a tab is being managed by another addon and that the user doesn't want it to be managed by tridacytl?

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+import { browserBg, activeTabId } from "./webext"
+
+/**
+ * Return the active window's id.
+ *
+ */
+export async function activeWindowId() {
+    return (await browserBg.windows.getCurrent()).id
+}
```
I think it might be better to put this function near activeTabId, since they're similar.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ *
+ */
+export async function tabTgroup(id?: number) {
+    return browserBg.sessions.getTabValue(id || await activeTabId(), "tridactyl-tgroup")
```
I'm not seeing any restrictions on tab IDs on MDN so I think `id` could be `0` here, which would result in unexpected behavior.

noctuid replied:
I always forget 0 is falsey in some languages. Even if it couldn't be 0, I think an explicit check is better; updated.

glacambre commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ */
+export async function clearTabTgroup(id?: number | number[]) {
+    const ids = await tabIdsOrCurrent(id)
+    return Promise.all(ids.map(async (id) => {
+        await browserBg.sessions.removeTabValue(id, "tridactyl-tgroup")
+    }));
```
Better to perform the removals in parallel:
```suggestion
    return Promise.all(ids.map(id => browserBg.sessions.removeTabValue(id, "tridactyl-tgroup")))
        .then(() => undefined);
```

bovine3dom commented:
printWidth is only disabled for excmds.ts to stop it breaking our python macros: https://github.com/tridactyl/tridactyl/blob/80113f26ddf8d9cfea45b3539d6ea0eddcd41600/.prettierrc#L14 - function signatures must be on a single line.

It's enabled for everything else except comments, which should not have hard line wrapping as we assume everyone has an editor capable of soft line wrapping. I don't really care too much about this though. The default is 80 characters if you want something to aim for for function bodies.

`config.get` should be used almost all the time as it's much faster. Only use `config.getAsync` in places where code will be executed very early in the life of a tab where `config.get` may not have been populated by the user configuration yet.

I'll try to answer your other questions later today if glacambre doesn't beat me to it. Please remind me if I forget!

noctuid commented on src/excmds.ts:
```diff
@@ -2664,7 +2666,191 @@ export async function viewcontainers() {
+/** @hidden */
+//#content_helper
+// {
+browser.tabs.onCreated.addListener(tgroupHandleTabCreated)
+browser.tabs.onRemoved.addListener(tgroupHandleTabRemoved)
+browser.tabs.onDetached.addListener(tgroupHandleTabDetached)
+browser.tabs.onAttached.addListener(tgroupHandleTabAttached)
+browser.tabs.onActivated.addListener(tgroupHandleTabActivated)
+browser.tabs.onUpdated.addListener(tgroupHandleTabUpdated)
```
For whatever reason it does work. I've moved it to the background though.

noctuid commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+ * @param other Whether to run fun on tabs not in the tab group instead.
+ *
+ */
+export async function tgroupRun(name: string, other: boolean = false): Promise<browser.tabs.Tab[]> {
```
It used to (it also used to do what the docstring says it does), but I've refactored everything a billion times. Thanks for catching my mistake and for the detailed review.

noctuid commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,330 @@
+export async function tgroupRun(name: string, other: boolean = false): Promise<browser.tabs.Tab[]> {
+    return browserBg.tabs.query({currentWindow : true}).then(async (tabs) => {
+        const ids = tabs.map(tab => tab.id)
+        const sameGroupIndices = await Promise.all(ids.map(async (id) => {
```
Thanks, I forgot you could destructure things like this.

noctuid commented:
> Thanks a lot for this PR, it looks architecturally sound and pretty clean :).

Thanks for all the feedback; it's a lot better after those changes. You should have seen it before the first commit. You would have cried.

> Also, while I'm a semicolon person myself, I believe bovine3dom doesn't like them and it's better to have a consistent codebase, so these should be removed (I noticed some in tab_groups.ts.

I didn't even know it was a thing to not have them before seeing this codebase (I don't like them). Removed the ones I saw in `tab_groups.ts`.

> It's enabled for everything else except comments

Not sure if this is expected, but `yarn pretty` did nothing for me on `tab_groups.ts`. I've formatted it with prettier myself. I am now getting a code duplication lint warning, but it's just one function call split across three lines.

Also, is it okay if I add an .editorconfig file?

bovine3dom commented on src/excmds.ts:
```diff
@@ -2664,7 +2665,192 @@ export async function viewcontainers() {
+// }
+
+/**
+ * Create a new tab group in the current window.
```
I think we should explain what tab groups are here for people who are unfamiliar with the concept.

bovine3dom commented on src/lib/config.ts:
```diff
@@ -928,6 +928,16 @@ export class default_config {
+     * Whether :tab shows completions for hidden tabs (e.g. tabs in other tab groups).
+     */
+    tabshowhidden: "true" | "false" = "false"
+
```
Please add `exaliases` for `tgroup* -> tabgroup*` to make them more discoverable when people type `:tab`. You will have to add each one individually.

bovine3dom commented on src/manifest.json:
```diff
@@ -53,6 +53,7 @@
         "search",
         "sessions",
         "storage",
+        "tabHide",
```
Please add a line to `doc/amo.md` explaining why we need this permission :)

bovine3dom commented:
I've added the `blocked` label as this requires a new permission which means that we can expect it to take months for users to update to a version containing this patchset. Related: #788.

I'd like to fix #2329 and some other P1/P2 bugs and get that patch out to users first.

That probably means I won't merge this for a few weeks (at best). I'd really like to see `tgroupswitch` completions if you're up for that, @noctuid.

Alternatively, there was talk of optional permissions being useful now, so perhaps you can make use of those and avoid #788? https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions - unfortunately, we can't ask for permissions outside of "user actions", which aren't fired by the command line, so we'd need to add e.g. a `:permissions` page that lets people click buttons to grant permissions - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/request.

bovine3dom commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,379 @@
+ *
+ */
+export async function clearTgroups() {
+    return browserBg.sessions.removeWindowValue(
```
We could cheer up tslint with

```
#src/lib/webext.ts
export removeActiveWindowValue = async value => browserBg.sessions.removeWindowValue(
    await activeWindowId(),
    value,
)

#src/lib/tab_groups.ts
import { activeTabId, activeWindowId, removeActiveWindowValue, browserBg } from "./webext"
clearTgroups = () => removeActiveWindowValue("tridactyl-tgroups")
```

noctuid commented on src/lib/tab_groups.ts:
```diff
@@ -0,0 +1,379 @@
+ *
+ */
+export async function tgroupHandleTabCreated(tab: browser.tabs.Tab) {
+    const windowGroup = await windowTgroup()
+
+    if (windowGroup) {
+        const tabGroup = await tabTgroup(tab.id)
+        if (!tabGroup) {
+            return setTabTgroup(windowGroup, tab.id)
+        }
+    }
```
Yeah it would. I didn't realize that was possible. How can I create a tab in another window like this?

noctuid commented:
> There is no good way to do this. We should probably ask Mozilla for tabValueChanged/windowValueChanged events. In the mean time, you can probably add a dummy field to the tab State structure and touch it when commands such as tgroupcreate and tgrouprename are used.

Thanks, will look into doing this later.

> I think this is because git didn't know about this file before you added it.

I had already added it in a previous commit. It looks like there other files where the printWidth isn't 80 too.    

BTW, given that tslint is deprecated, are there any plans to switch to eslint?  

> That probably means I won't merge this for a few weeks (at best).

Fine by me. I think it would be preferable for it to just work without having to go in and manually enable the permission.

> I'd really like to see tgroupswitch completions if you're up for that,

I added a file for it but haven't gotten a chance to figure out how the completion system actually works yet. I will look into it.

bovine3dom commented:
Plan for this: I'd like to fix #1409 and get it released before merging this. If I can't do that within a fortnight I'll merge this anyway.

noctuid commented:
Don't merge before I've fixed the group display in the mode indicator and fixed completion. I will try to do this in the next couple of weeks.

noctuid commented:
Completion works now.

> In the mean time, you can probably add a dummy field to the tab State structure and touch it when commands such as tgroupcreate and tgrouprename are used.

I added a `group` field, but if I try to set `contentState.group` from within tgroupcreate or tgrouprename, nothing happens. If I try to create a function in #content that does this and call it, I get an error. What's the correct way to do this?

> Won't this have surprising results if a tab is created in a window that isn't active? Perhaps this should use the tab group of the window the tab is created in?

> Yeah it would. I didn't realize that was possible. How can I create a tab in another window like this?

@glacambre Can you elaborate on this comment?

glacambre commented:
On 2020-05-30 04:11, Fox Kiester wrote:
> Completion works now.

Nice!

>>In the mean time, you can probably add a dummy field to the tab
>>State structure and touch it when commands such as tgroupcreate and
>>tgrouprename are used.
> 
> I added a |group| field, but if I try to set |contentState.group| from 
> within tgroupcreate or tgrouprename, nothing happens. If I try to create 
> a function in #content that does this and call it, I get an error. 
> What's the correct way to do this?

You could try to make the function a #background_helper (or maybe it 
doesn't actually need to be marked with #{content,background}? I don't 
remember exactly how all of this works).

>>>Won't this have surprising results if a tab is created in a window
>>>that isn't active? Perhaps this should use the tab group of the
>>>window the tab is created in?
>>
>>Yeah it would. I didn't realize that was possible. How can I create
>>a tab in another window like this?
> 
> @glacambre <https://github.com/glacambre> Can you elaborate on this comment?

Sorry - I forgot to reply :). You can create a tab in another window 
with `:jsb browser.tabs.create({ active: false, windowId: 1 })`. Use `js 
tri.browserBg.windows.getAll().then(r => alert(r.map(w => w.id).join(", 
")))` to get a list of window IDs.

noctuid commented:
> Sorry - I forgot to reply :). You can create a tab in another window
with `:jsb browser.tabs.create({ active: false, windowId: 1 })`. Use `js
tri.browserBg.windows.getAll().then(r => alert(r.map(w => w.id).join(",
")))` to get a list of window IDs.

Thanks, I updated it so that none of the event listeners use the current window ID. Tested with those commands.

> You could try to make the function a #background_helper (or maybe it
doesn't actually need to be marked with #{content,background}? I don't
remember exactly how all of this works).

It doesn't seem to matter. Trying to set `contentState.group` (or even `contentState.mode`) anywhere doesn't cause the status indicator to update. Do you have any other ideas?

bovine3dom commented on src/excmds.ts:
```diff
@@ -2664,7 +2665,209 @@ export async function viewcontainers() {
+/** @hidden */
+//#content
+function setContentStateGroup(name: string) {
+    contentState.group = name
```
I think this should work. I'll look into it today.

bovine3dom commented on src/excmds.ts:
```diff
@@ -2664,7 +2665,209 @@ export async function viewcontainers() {
+
+/** @hidden */
+//#content
+function setContentStateGroup(name: string) {
```
```suggestion
export function setContentStateGroup(name: string) {
```

Content functions need to be exported to be called from the background script : )

bovine3dom commented:
With the fix I've suggested, if I do `tgroupcreate blah` the mode indicator shows `blah | blah` until I run any other Tridactyl command (e.g. `j` to scroll down).

This is probably because I was lazy when I made the mode indicator (if I even made it?) but it would be nice to get it fixed before we merge this.

noctuid commented:
> This is probably because I was lazy when I made the mode indicator

That was my fault. Fixed now. Rebased and squashed.

There are still many situations where the status line does not update immediately. Some of them I'm not sure how to handle. For example, after `tgroupabort`. I can call `setContentStateGroup`, but that only has an effect on the current tab. All background tabs will still initially show the old tab group. Would it be a bad idea to call `setContentStateGroup` on tab switch? Is there some way I could run this for all tabs after `tgroupabort` instead?

bovine3dom commented:
Sorry that this is taking so long for me to get round to. I'm going to try very hard to look at it tomorrow.

bovine3dom commented:
> I can call setContentStateGroup, but that only has an effect on the current tab. All background tabs will still initially show the old tab group. Would it be a bad idea to call setContentStateGroup on tab switch? Is there some way I could run this for all tabs after tgroupabort instead?

We probably should just try to fix the mode indicator in general (e.g. have an "update mode indicator on all tabs" command). So I wouldn't worry too much about it.

This PR seems to work fine and looks fine to me. So it's only blocked because it needs more permissions. I'll merge everything in the "more-permissions" milestone at the same time.

Thanks for this and sorry it took so long for me to get round to looking at it!

dlobue commented:
Any ETA on getting this merged in? I'd LOVE to start using it!

bovine3dom commented:
I had completely forgotten about this PR.

I've given it a "blocks next release" tag which is a lie but it should mean that I see it when I next try to do a release.

Thanks for bumping it!
